### ISSUE TITLE
Regression after stockfish-nnue-2020-07-19

### ISSUE THREAD

**Issue by nodchip** [nodchip/Stockfish#141]
It is reported that there was regression between https://github.com/nodchip/Stockfish/releases/tag/stockfish-nnue-2020-07-19 and https://github.com/nodchip/Stockfish/releases/tag/stockfish-nnue-2020-08-11.  https://discordapp.com/channels/435943710472011776/737073080081317948/754020352589561876

We could need to investigate the root cause of the regression.

**Comment by vondele** [comment]
I have approved the corresponding fishtest tests. 

I'm wondering if the issue is player or trainer related, fishtest will only pick up the former, I would think?

**Comment by nodchip** [comment]
To investigate this issue.  I planned to generate training data with a commit, validation data with the commit, train nets with several commits, compare elos of those nets.

At first, I created training data with https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba and the following commands.
```
uci
setoption name Use NNUE value false
setoption name Threads value 27
setoption name Hash value 32768
setoption name SyzygyPath value D:\data\chess\tablebase\3-4-5
isready
gensfen depth 6 loop 100000000 use_draw_in_training_data_generation 1 eval_limit 32000 output_file_name F:\stockfish\training_data\2020-09-12\Threads=27.Hash=32768.depth=6.loop=100000000.use_draw_in_training_data_generation=1.eval_limit=32000.bin
quit
``` 

Then, I created validation data with https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba and the following commands.
```
uci
setoption name Use NNUE value false
setoption name Threads value 27
setoption name Hash value 32768
setoption name SyzygyPath value D:\data\chess\tablebase\3-4-5
isready
gensfen depth 6 loop 1000000 use_draw_in_training_data_generation 1 eval_limit 32000 output_file_name F:\stockfish\validation_data\2020-09-12\generated_kifu.bin
quit
```

Then, I trained nets with the following commits and options to investigate this issue.

1. https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (=https://github.com/nodchip/Stockfish/releases/tag/stockfish-nnue-2020-07-19)
``` 
uci
setoption name SkipLoadingEval value true
setoption name Threads value 16
setoption name EvalSaveDir value F:\stockfish\eval\2020-07-19.36092b855a5e2bcfb587a36e2055ea068e4bd8e5
isready
learn targetdir F:\stockfish\training_data\2020-09-12 loop 100 batchsize 1000000 use_draw_in_training 1 use_draw_in_validation 1 eta 1 lambda 1 eval_limit 32000 nn_batch_size 1000 newbob_decay 0.5 eval_save_interval 100000000 loss_output_interval 1000000 mirror_percentage 0 validation_set_file_name F:\stockfish\validation_data\2020-09-12\generated_kifu.bin
quit
```

2. https://github.com/nodchip/Stockfish/commit/35f04aaf24d05ec7b71c2cc1710d00642aaf35ec (=https://github.com/nodchip/Stockfish/releases/tag/stockfish-nnue-2020-08-11)
```
uci
setoption name SkipLoadingEval value true
setoption name Use NNUE value true
setoption name Threads value 16
setoption name EvalSaveDir value F:\stockfish\eval\2020-08-11.35f04aaf24d05ec7b71c2cc1710d00642aaf35ec
isready
learn targetdir F:\stockfish\training_data\2020-09-12 loop 100 batchsize 1000000 use_draw_in_training 1 use_draw_in_validation 1 eta 1 lambda 1 eval_limit 32000 nn_batch_size 1000 newbob_decay 0.5 eval_save_interval 100000000 loss_output_interval 1000000 mirror_percentage 0 validation_set_file_name F:\stockfish\validation_data\2020-09-12\generated_kifu.bin
quit
```

3. https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec (=https://github.com/nodchip/Stockfish/releases/tag/stockfish-nnue-2020-08-30)
```
uci
setoption name SkipLoadingEval value true
setoption name Use NNUE value true
setoption name Threads value 16
setoption name EvalSaveDir value F:\stockfish\eval\2020-08-30.bc90567e0933a49f8c1911eba45ec6743a37c1ec
isready
learn targetdir F:\stockfish\training_data\2020-09-12 loop 100 batchsize 1000000 use_draw_in_training 1 use_draw_in_validation 1 eta 1 lambda 1 eval_limit 32000 nn_batch_size 1000 newbob_decay 0.5 eval_save_interval 100000000 loss_output_interval 1000000 mirror_percentage 0 validation_set_file_name F:\stockfish\validation_data\2020-09-12\generated_kifu.bin use_raw_nnue_eval 1
quit
```

4. https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba with EnableTranspositionTable=true
```
uci
setoption name SkipLoadingEval value true
setoption name Use NNUE value pure
setoption name Threads value 16
setoption name EnableTranspositionTable value true
setoption name EvalSaveDir value F:\stockfish\eval\2020-09-12.bcfe28b2ae468d045e2f96b659401422531bacba.EnableTranspositionTable=true
isready
learn targetdir F:\stockfish\training_data\2020-09-12 loop 100 batchsize 1000000 use_draw_in_training 1 use_draw_in_validation 1 eta 1 lambda 1 eval_limit 32000 nn_batch_size 1000 newbob_decay 0.5 eval_save_interval 100000000 loss_output_interval 1000000 mirror_percentage 0 validation_set_file_name F:\stockfish\validation_data\2020-09-12\generated_kifu.bin
quit
```
 
5. https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba with EnableTranspositionTable=false
```
uci
setoption name SkipLoadingEval value true
setoption name Use NNUE value pure
setoption name Threads value 16
setoption name EnableTranspositionTable value false
setoption name EvalSaveDir value F:\stockfish\eval\2020-09-12.bcfe28b2ae468d045e2f96b659401422531bacba.EnableTranspositionTable=false
isready
learn targetdir F:\stockfish\training_data\2020-09-12 loop 100 batchsize 1000000 use_draw_in_training 1 use_draw_in_validation 1 eta 1 lambda 1 eval_limit 32000 nn_batch_size 1000 newbob_decay 0.5 eval_save_interval 100000000 loss_output_interval 1000000 mirror_percentage 0 validation_set_file_name F:\stockfish\validation_data\2020-09-12\generated_kifu.bin
quit
```
The results of the trainings is below.

![image 2020-09-13-20-20-13](https://user-images.githubusercontent.com/685625/93016800-c4c58a00-f5fe-11ea-8fac-852a20df3cd8.png)

The loss of https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba was not decreased, and the output net file was identical to the initial random network parameters.

We can see several facts from the result.
- The test cross entropies of https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5, https://github.com/nodchip/Stockfish/commit/35f04aaf24d05ec7b71c2cc1710d00642aaf35ec and https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec are different from https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba.  https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5, https://github.com/nodchip/Stockfish/commit/35f04aaf24d05ec7b71c2cc1710d00642aaf35ec and https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec look good, and https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba looks bad.
- The move accuracy of https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 is better than https://github.com/nodchip/Stockfish/commit/35f04aaf24d05ec7b71c2cc1710d00642aaf35ec and https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec.
- The move accuracies of https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba with EnableTranspositionTable=true and https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba with EnableTranspositionTable=false are different.

We can think there are two regressions.
- There was a regression in the trainer between https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (=stockfish-nnue-2020-07-19) and https://github.com/nodchip/Stockfish/commit/35f04aaf24d05ec7b71c2cc1710d00642aaf35ec (=stockfish-nnue-2020-08-11).
- There was another regression between https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec (=https://github.com/nodchip/Stockfish/releases/tag/stockfish-nnue-2020-08-30) and https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba.

To comfirm that the first regression is an actual regression, I posted three tests to fishtest.
- https://tests.stockfishchess.org/tests/view/5f5df39461bdd7a0007e76d9
- https://tests.stockfishchess.org/tests/view/5f5df3d461bdd7a0007e76db
- https://tests.stockfishchess.org/tests/view/5f5df3fd61bdd7a0007e76dd

I will wait for those results.

**Comment by vondele** [comment]
@nodchip tests failed to start as the bench didn't match (see also https://tests.stockfishchess.org/actions)

**Comment by nodchip** [comment]
@vondele Thank you for notification.  It should be because the net is not the default one.  I will purge the current tests, and submit new tests.

**Comment by nodchip** [comment]
I submitted new tests.

- https://tests.stockfishchess.org/tests/view/5f5e0d2861bdd7a0007e76fa
- https://tests.stockfishchess.org/tests/view/5f5e0d7e61bdd7a0007e76fc
- https://tests.stockfishchess.org/tests/view/5f5e0db661bdd7a0007e76fe

**Comment by nodchip** [comment]
https://github.com/nodchip/Stockfish/commit/35f04aaf24d05ec7b71c2cc1710d00642aaf35ec (2020-08-11) vs https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (2020-07-19)
ELO: -6.79 +-1.7 (95%) LOS: 0.0%
Total: 60000 W: 11291 L: 12464 D: 36245
Ptnml(0-2): 1254, 7452, 13560, 6681, 1053

https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec (2020-08-30) vs https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (2020-07-19)
ELO: -28.50 +-1.8 (95%) LOS: 0.0%
Total: 60000 W: 9834 L: 14744 D: 35422
Ptnml(0-2): 1718, 8785, 12979, 5725, 793

https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec (2020-08-30) vs https://github.com/nodchip/Stockfish/commit/35f04aaf24d05ec7b71c2cc1710d00642aaf35ec (2020-08-11)
ELO: -19.52 +-1.8 (95%) LOS: 0.0%
Total: 60000 W: 10763 L: 14131 D: 35106
Ptnml(0-2): 1648, 8136, 13127, 6114, 975

The strength of the nets are https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (2020-07-19) > https://github.com/nodchip/Stockfish/commit/35f04aaf24d05ec7b71c2cc1710d00642aaf35ec (2020-08-11) > https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec (2020-08-30).  There would be some regresisons between https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec (2020-08-30) and https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (2020-07-19).

There at lease two regressions.

- Between https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec (2020-08-30) and https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (2020-07-19).
- After https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec (2020-08-30).

I will split this issue into two issue, https://github.com/nodchip/Stockfish/issues/151 and https://github.com/nodchip/Stockfish/issues/152, and will continue to investigate.  I will close this issue for now.
